[이지예] 4주차 크리스마스 문제 풀이

jiyeeeah opened the pull request:
## 🎄 크리스마스 이벤트 기능 구현 목록

### 📌 날짜와 메뉴 입력

- [x] 식당 예상 방문 날짜 입력 받기 (`12월 중 식당 예상 방문 날짜는 언제인가요? (숫자만 입력해 주세요!)`)
  - 1 이상 31 이하의 숫자만 가능 (`[ERROR] 유효하지 않은 날짜입니다. 다시 입력해주세요.`)
- [x] 주문할 메뉴와 개수 입력 받기 (`주문하실 메뉴와 개수를 알려 주세요. (e.g. 해산물파스타-2,레드와인-1,초코케이크-1)`)
  - 메뉴판에 없는 메뉴 (`[ERROR] 유효하지 않은 주문입니다. 다시 입력해 주세요.`)
  - 메뉴의 개수 1개 이상의 숫자만 가능 (`[ERROR] 유효하지 않은 주문입니다. 다시 입력해 주세요.`)
  - 메뉴 형식이 예시와 다른 경우 (`[ERROR] 유효하지 않은 주문입니다. 다시 입력해 주세요.`)
  - 중복 메뉴 (`[ERROR] 유효하지 않은 주문입니다. 다시 입력해 주세요.`)
  - 음료만 주문시, 주문할 수 없음
  - 메뉴는 한번에 최대 20개까지만 가능
- [x] 혜택 미리보기! 문구 출력 (`12월 00일에 우테코 식당에서 받을 이벤트 혜택 미리 보기!`)
- [x] 입력받은 메뉴 바탕으로 주문 메뉴 출력

### 📌 이벤트 혜택 미리보기

- [x] 할인 전 총 주문 금액 출력
- [x] 증정 이벤트 출력
  - 할인 전 총 주문 금액이 12만원 이상이면 `샴페인 1개` 증정
- [x] 혜택 내역 출력
  - 적용된 할인 이벤트 및 혜택 금액 출력
- [x] 총 혜택 금액 출력
  - 할인 금액의 합계 + 증정 메뉴의 가격
- [x] 할인 후 예상 결제 금액 출력
  - 할인 전 총 주문 금액 - 할인 금액
- [x] 12월 이벤트 배지 출력
  - 총 혜택 금액에 따라 이벤트 배지 부여

### 프로그래밍 요구 사항

- indent
- Jest
- 함수 길이 15라인 넘지 않도록, 최대한 작게
- else 지양
- 예외 발생 이후 해당 부분부터 입력 다시 받기
- InputView, OutputView 객체를 활용해 구현

---
### 소감...

[블로그](https://velog.io/@jiyeeeah/6%EA%B8%B0-%ED%94%84%EB%A6%AC%EC%BD%94%EC%8A%A4-4%EC%A3%BC%EC%B0%A8-%EB%AC%B8%EC%A0%9C%ED%92%80%EC%9D%B4)에 소감(?)같은걸 작성해봤습니다...
이번엔 여러모로 아쉬움이 많네요ㅠ 테스트 코드도 작성하지 못했고 뭔가 구조가 정돈되지 못한 것 같습니다.. 아쉽...

jiyeeeah committed:
[docs] 기능 구현 목록 작성

jiyeeeah committed:
[feat] 입출력 메시지, 에러메시지, 메뉴 상수화

jiyeeeah committed:
[feat] 식당 예상 방문 날짜 입력 받기

- App.js의 getDate 메서드에서 InputView를 사용해 입력값을 받고
  Validation을 통해 유효성 검사를 거칩니다.
- this.date에 방문 날짜를 저장합니다.

jiyeeeah committed:
[feat] 주문 메뉴와 개수 입력받기

- App.js의 getMenus 메서드에서 InputView를 활용해 메뉴를 입력받고
  Validation을 통해 유효성 검사를 함
- this.menus에 입력받은 메뉴 저장

jiyeeeah committed:
[docs] 추가 기능 목록 작성

- 고객에게 안내되는 이벤트 주의사항 추가

jiyeeeah committed:
[refactor] 메뉴 상수 카테고리화

- 메뉴 담고있던 상수를 카테고리화함
- 그에 맞게 validation 코드 수정

jiyeeeah committed:
[feat] 추가된 기능 요구 사항에 따른 유효성 검사 추가

- 음료만 주문시, 또는 총 메뉴 개수가 20개 초과시 예외 발생

jiyeeeah committed:
[feat] 혜택 미리보기! 문구 출력

jiyeeeah committed:
[feat] 주문 받은 메뉴 출력

- App.js에서 menu를 받을 때 this.menus에 Map의 형태로 저장하도록 수정
- OutputView에서 Map형태의 메뉴목록을 바탕으로 주문받은 메뉴 출력

jiyeeeah committed:
[refactor] menu 상수 변경

jiyeeeah committed:
[refactor] App.js 내에서 사용되는 필드들 private으로 선언

- App.js 내에서 사용되는 필드들이 숨어있어 찾기 어려움
- 최상단에서 미리 private으로 선언하여 App class안에서만 사용할 수 있도록 함

jiyeeeah committed:
[feat] 할인 전 총 주문 금액 출력

- App.js에서 getTotalAmount를 통해 총 주문 금액 계산
- OutputView에서 총 주문 금액 출력

jiyeeeah committed:
[feat] 증정 이벤트 출력

- App.js의 getOffer 메서드를 통해 증정 여부를 boolean으로 return
- OutputView에서 증정여부에 따라 출력

jiyeeeah committed:
[refactor] 증정 메뉴 상수화

- 증정 메뉴가 바뀌어도 유지보수가 쉽게 하기 위해,
  menu.js 파일 내에 BENEFIT_MENU 객체에 증정 메뉴와 가격 저장해둠

jiyeeeah committed:
[feat] 혜택 내역 출력

- App.js의 discount라는 객체에 할인 여부 계산해서 넣음

jiyeeeah committed:
[feat] 총 혜택 금액 출력

jiyeeeah committed:
[feat] 할인 후 예상 결제 금액 출력

jiyeeeah committed:
[feat] 12월 이벤트 배지 출력

jiyeeeah committed:
[refactor] event 관련 내용 상수화

- event 출력 string과 계산 및 할인 가격을 상수화함
- BENEFIT_MENU -> OFFER_MENU 네이밍 변경 및 위치 event.js 파일로 이동

jiyeeeah committed:
[refactor] 이벤트 배지 상수화

- 12월 이벤트 배지의 이름과 기준 가격을 상수화함

jiyeeeah committed:
[refactor] EVENT 상수 사용

jiyeeeah committed:
[refactor] EVENT 적용되는 날들 상수화

- Weekdays는 Weekends를 제외한 모든 날임

jiyeeeah committed:
[refactor] 증정 이벤트 기준도 상수에 포함시킴

jiyeeeah committed:
[docs] 각 상수, 필드, 메서드, 함수에 주석 추가

jiyeeeah committed:
[refactor] 일부 메서드명 수정

jiyeeeah committed:
[refactor] 증정 여부 discount 필드 객체 안으로 옮김

jiyeeeah committed:
[refactor] 최대한 메서드 내에서 필드 바꾸지 않도록 수정
- 메서드들은 값을 return하고 run() 에서 필드에 return 값들을 저장하도록 함
- 단, discount객체 제외

jiyeeeah committed:
[docs] claculateTotalAmount 내 주석 추가

wo-o29 commented on src/App.js:
```diff
@@ -1,5 +1,135 @@
+  /** 주문 받은 메뉴를 기반으로 할인 전 총 주문금액을 구합니다. */
+  #calculateTotalAmount() {
+    // 메뉴판의 모든 메뉴와 가격을 객체 형태로 새롭게 생성
+    const allMenuPrice = {};
+    Object.values(MENU).forEach((category) =>
+      Object.entries(category).forEach(([_i, { name, price }]) => (allMenuPrice[name] = price))
+    );
```
여기 로직은 전체 메뉴를 `메뉴: 가격`형태의 객체로 만드는 것 같아서 조금 더 범용적으로 사용할 수 있는 것 같은데 분리해보는건 어떨까요??

jiyeeeah replied:
그럴까요? 고민해보고 리팩토링 해보겠습니다! 감사합니다 :)

wo-o29 commented on src/App.js:
```diff
@@ -1,5 +1,135 @@
+      Object.entries(category).forEach(([_i, { name, price }]) => (allMenuPrice[name] = price))
+    );
+
+    let totalAmount = 0;
+    this.#menusMap.forEach((count, dish) => {
+      totalAmount += allMenuPrice[dish] * count;
+    });
+    return totalAmount;
```
`reduce` 한번 검색해보시고 사용해보시는 것도 좋을 것 같아요! 

물론 `forEach`로 결과를 만들 수 있지만 지금 목적에서는 `reduce`가 조금 더 어울리는 것 같네요🙃

jiyeeeah replied:
아직 그런 메서드 사용이 어색하네요ㅠㅠ 존재를 알고 있어도 막상 사용하려면 헷갈리는.. 좀더 연습해보겠습니다! 감사합니다 👍

jiyeeeah replied:
여기선 `Map`객체여서 `reduce`사용이 어렵겠네요ㅠㅠ 그래도 좋은 조언 감사합니당!

jiyeeeah replied:
```suggestion
    return Array.from(menusMap.entries()).reduce((acc, [dish, count]) => {
      return acc + allMenuPrice[dish] * count;
    }, 0);
```

이렇게는 수정 가능한 것 같긴 합니다! 👍

wo-o29 commented on src/App.js:
```diff
@@ -1,5 +1,135 @@
+  #isOffer() {
+    if (this.#totalAmount >= OFFER_MENU.applyPoint) return true;
+    return false;
+  }
```
```suggestion
  #isOffer() {
    return this.#totalAmount >= OFFER_MENU.applyPoint;
  }
```
이렇게 생략할 수 있을 것 같아요!

wo-o29 commented on src/constants/event.js:
```diff
@@ -0,0 +1,37 @@
+export const OFFER_MENU = Object.freeze({ string: "샴페인 1개", price: 25000, applyPoint: 120000 });
+
+/** 이벤트 종류와 그에 대한 출력 문자열, 계산법 */
+export const EVENT = Object.freeze({
+  CHRISTMAS: {
+    string: (amount) => `크리스마스 디데이 할인: -${amount.toLocaleString()}원`,
+    calculate: (day) => 1000 + (day - 1) * 100,
+  },
+  WEEKDAYS: {
+    string: (amount) => `평일 할인: -${amount.toLocaleString()}원`,
+    calculate: (count) => 2023 * count,
+  },
+  WEEKENDS: {
+    string: (amount) => `주말 할인: -${amount.toLocaleString()}원`,
+    calculate: (count) => 2023 * count,
+  },
+  SPECIAL: { string: `특별 할인: -1,000원`, calculate: 1000 },
+  OFFER: {
+    string: `증정 이벤트: -${OFFER_MENU.price.toLocaleString()}원`,
+    calculate: OFFER_MENU.price,
+  },
+});
```
저는 이정도면 `utils`에 가깝다고 생각하는데 상수로 생각하신 이유가 있으신가요??

jiyeeeah replied:
상수라기보다는 뭔가 이벤트 관련된 룰?을 저장해놓은 객체라고 생각했어요! `utils`로 보내는게 맞을까요..? 
제가 `constants`에 넣은 것들은 상수보다는 게임 전체적인 룰과 인터페이스 관련된 내용이 맞는 것 같아요..! 폴더명을 `rules`로 바꿔야하나 🤔

wo-o29 replied:
```js
WEEKENDS: {
    string: (amount) => `주말 할인: -${amount.toLocaleString()}원`,
    calculate: (count) => 2023 * count,
  },
```
이런 부분은 `utils`에 가깝다고 생각하는데 `SPECIAL: { string: 특별 할인: -1,000원, calculate: 1000 },` 이런건 또 `constants`에 맞는 것 같아서 어려운 것 같아요🤣

jiyeeeah replied:
마자요... 일단 고려해서 리팩토링 해보겠습니다! 좋은 관점 제시해주셔서 감사해용!

wo-o29 commented on src/App.js:
```diff
@@ -1,5 +1,135 @@
+      this.#discount.christmas = EVENT.CHRISTMAS.calculate(this.#date);
+    }
+
+    if (EVENT_DAYS.WEEKENDS.includes(this.#date)) {
+      const mains = MENU.main.map((item) => item.name);
+      const count = this.#calculateWeekDiscount(mains);
+      this.#discount.weekends = EVENT.WEEKENDS.calculate(count);
+    } else {
+      const desserts = MENU.dessert.map((item) => item.name);
+      const count = this.#calculateWeekDiscount(desserts);
+      this.#discount.weekdays = EVENT.WEEKDAYS.calculate(count);
+      if (EVENT_DAYS.SPECIAL.includes(this.#date)) this.#discount.special = true;
+    }
```
```suggestion
    if (EVENT_DAYS.WEEKENDS.includes(this.#date)) {
      const mains = MENU.main.map((item) => item.name);
      const count = this.#calculateWeekDiscount(mains);
      this.#discount.weekends = EVENT.WEEKENDS.calculate(count);
      return;
    }
    
      const desserts = MENU.dessert.map((item) => item.name);
      const count = this.#calculateWeekDiscount(desserts);
      this.#discount.weekdays = EVENT.WEEKDAYS.calculate(count);
      
      if (EVENT_DAYS.SPECIAL.includes(this.#date)) this.#discount.special = true;
```
조건이 너무 많고 해서 가독성이 떨어지는 것 같네요... `early return`을 사용해봐도 좋을 것 같아요!

jiyeeeah replied:
오호! 이 경우에는 `early return`이 가독성에 도움이 되는 군요! 헷갈리던 부분이었는데 감사합니다 ☺️

wo-o29 commented on src/App.js:
```diff
@@ -1,5 +1,135 @@
+      if (targetMenus.includes(dish)) count += num;
+    }
+    return count;
+  }
```
여기도 `reduce`로 해결할 수 있을 것 같습니다~!

wo-o29 commented on src/OutputView.js:
```diff
@@ -1,7 +1,102 @@
+  printOffer(isOffer) {
+    Console.print(OUTPUT_MESSAGES.EVENT_OFFER);
+    if (isOffer) Console.print(OFFER_MENU.string);
+    else Console.print("없음");
```
early return을 사용하면 `else`문을 지양할 수 있을 것 같아요!

jiyeeeah replied:
저는 여기서는 `else`문을 사용하는게 맞다고 생각이 들었어요! `early return`을 쓰는 것보다 이게 더 가독성이 좋아보인다고 생각했습니다 :)

wo-o29 commented on src/OutputView.js:
```diff
@@ -1,7 +1,102 @@
+   * @param {Object} discount
+   * @param {boolean} isOffer
+   */
+  printDiscount(discount, isOffer) {
```
아마 느끼셨겠지만 여기 부분이 유지보수성도 떨어지고 비효율적인 것 같아서 개선해볼 수 있는 방법을 고민해보셔도 좋을 것 같아요🤔

jiyeeeah replied:
ㅎㅎ 맞아요.. 더 고민이 필요한 부분입니다 😇

wo-o29 commented on src/OutputView.js:
```diff
@@ -1,7 +1,102 @@
+      isNoBenefit = false;
+    }
+    if (discount.special) {
+      Console.print(EVENT.SPECIAL.string);
```
```suggestion
      Console.print(EVENT.SPECIAL.STRING);
```
상수라면 내부  속성도 스네이크 케이스를 사용하는건 어떨까요??

wo-o29 commented on src/validation.js:
```diff
@@ -0,0 +1,39 @@
+    if (![...menuSet].some((dish) => !baverages.includes(dish)))
+      throw new Error(ERROR_MESSAGES.INVALID_MENU);
+    if (menuCount > 20) throw new Error(ERROR_MESSAGES.INVALID_MENU);
+  },
```
음... 여기부분도 가독성이 떨어지는 것 같아서 개행을 추가하거나 구조를 분리하는 방식으로 개선해보는 것도 좋을 것 같아요🤔

wo-o29 reviewed (approved):
아무래도 외부에 의존하는 게 강하기 때문에 유닛 테스트가 어려우셨을 것 같네요...!
유닛 테스트할 수 있게 구조를 생각해보시는 것도 좋을 것 같아요!

그리고 이번 요구사항이 조건도 많고 복잡해서 전체적으로 가독성에 아쉬운 부분이 있어서 리팩토링때 이러한 부분도 개선해보면 좋을 것 같아요🫡

고생하셨습니다👍👏🎉🔥

jiyeeeah committed:
[feat] 이벤트 시작 문구 출력

ssilver01 commented on docs/README.md:
```diff
@@ -0,0 +1,40 @@
+## 🎄 크리스마스 이벤트 기능 구현 목록
```
기능 구현 목록을 너무 깔끔히 정리하신 것 같아요 ..👍

ssilver01 commented on src/validation.js:
```diff
@@ -0,0 +1,39 @@
+   */
+  menus(input) {
+    const menuSet = new Set();
+    const allDishes = Object.values(MENU).flatMap((category) => category.map((item) => item.name));
```
flatMap과 map의 차이점이 뭔가요?!

jiyeeeah replied:
`map()`은 요소 각각에 대하여 주어진 함수를 호출하여 전달받은 `return` 값을 모아 기존과는 다른 새로운 배열을 만들어내는 반면, 
`flatMap()`은 요소 각각에 대해 콜백함수에서 전달받은 `return` 값을 모아 **깊이를 `-1`한 후** 새로운 배열을 만들어냅니다!

간단한 예제는 다음과 같습니다!
```js
console.log([1, [3], [2]].map(ele => ele)) // [ 1, [ 3 ], [ 2 ] ]
console.log([1, [3], [2]].flatMap(ele => ele)) // [ 1, 3, 2 ]
```

[flatMap 문서](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/flatMap)
[map 문서](https://developer.mozilla.org/ko/docs/Web/JavaScript/Reference/Global_Objects/Map)

ssilver01 commented on src/constants/menu.js:
```diff
@@ -0,0 +1,23 @@
+    { name: "초코케이크", price: 15000 },
+    { name: "아이스크림", price: 5000 },
+  ],
+  baverage: [
```
오타가 있어요 beverage

jiyeeeah replied:
헉 그러네요! 감사합니다 :)

ssilver01 commented on src/validation.js:
```diff
@@ -0,0 +1,39 @@
+    const baverages = MENU.baverage.map((item) => item.name);
+    let menuCount = 0;
+
+    for (const menu of input) {
```
menus 유효성 검사 함수 분리하는건 어떨까요?!  그럼 가독성이 높아질 것 같아요 !

teagu123 replied:
저도 여기는 값을 props로 받더라도 함수로 나눠서 menus에서 불러오는게 더 가독성에 좋을거같습니다!!
어떤 validation인지 따로 주석이 없으면 보기가 조금 어려울거같습니다! ⭐️

jiyeeeah replied:
그렇군요!! 좋은 의견 감사합니다 ㅎㅎ 리팩토링하면서 수정해볼게요!

ssilver01 commented on src/InputView.js:
```diff
@@ -1,7 +1,23 @@
+   * @returns {string}
+   */
+  async readDate() {
+    const input = await Console.readLineAsync(INPUT_MESSAGES.VISIT_DATE);
+    return input;
+  },
```
줄이는게 가독성 측면에서 맞는지는 모르겠지만...
 return (await Console.readLineAsync(INPUT_MESSAGES.MENU)).split(",");
 이렇게도 줄여볼 수 있을 것 같아요 !

ssilver01 commented on src/App.js:
```diff
@@ -1,5 +1,136 @@
+  }
+
+  /** 날짜를 입력받고 this.#date에 날짜를 number형으로 저장합니다. */
+  async #getDate() {
+    try {
+      const date = await InputView.readDate();
+      Validation.date(date);
+      return Number(date);
+    } catch (error) {
+      Console.print(error.message);
+      await this.#getDate();
+    }
+  }
+
+  /** 주문할 메뉴를 입력받고 this.#menusMap에 메뉴를 Map<메뉴, 개수> 형식으로 저장합니다. */
+  async #getMenus() {
+    try {
+      const menus = await InputView.readMenus();
+      Validation.menus(menus);
+      return menus.map((menu) => menu.split("-"));
+    } catch (error) {
+      Console.print(error.message);
+      await this.#getMenus();
+    }
+  }
```
try catch를 사용했기에  await this.#getMenus();로 재귀호출을 하신이유가 있나요?!

jiyeeeah replied:
`getMenus`에서 에러가 생겼을 때 `getMenus()`를 다시 실행하도록 했습니다!

ssilver01 reviewed (commented):
LGTM 👍
커밋 내용이 정말 꼼꼼하시네요 ! 저도.. 그렇게 했어야됐는데,, 
정말 꼼꼼하신 것 같아요 ! 저번과 다르게 MVC를 적용하시지 않는 이유가 있으신가요?!

이번주도 수고하셨습니다⭐️

teagu123 commented on src/App.js:
```diff
@@ -1,5 +1,136 @@
+
+  async run() {
+    OutputView.printWelcome();
+    this.#date = await this.#getDate();
```
제가 잘몰라서 질문을 드립니다!
App.js를 보니 모든 return값들을 전역 private 필드에 넣으셔서 그거에 대한 질문입니다!
```suggestion
    const date= await this.#getDate();
```
이렇게 받지않으시고 #date (전역 private 변수)로 설정해서 받으시는 이유가 있으신가요?? 궁금합니다!

jiyeeeah replied:
이 부분은 저두 계속 고민중이라서 답변을 못드릴 것 같아요ㅠㅠ
이럴 때 private 필드를 사용하는게 맞는지 말씀하신대로 const로 선언해서 사용하는게 맞는건지...
아직 잘 모르겠습니다 ㅠㅠ

this.#date가 자주 사용되고 클래스 내에서 전역적으로 사용된다고 생각해서 필드 선언으로 하긴 했습니다..!
다만, 이렇게 하니 메서드들이 너무 의존성이 높아져서 리팩토링하면서 메서드들을 함수로 분리하고 있습니다! 파라미터로 받아서 return값을 내보내도록이요. 이렇게 하고나니 굳이 필드로 선언하지 않는게 더 유리해 보이긴 합니다...허허 혼란 투성이 ㅠ

teagu123 commented on src/OutputView.js:
```diff
@@ -1,7 +1,106 @@
+   */
+  printFinalAmount(amount) {
+    Console.print(OUTPUT_MESSAGES.TOTAL_AFTER_EVENT);
+    Console.print(`${amount.toLocaleString()}원`);
```
호호호 이런게 있군요!! 👍👍
저는 util에서 정규표현식을 써서 만들었다는....고생을 했네요 ㅎㅎ 저도 이걸로 변경해야겠군요

teagu123 commented on src/constants/event.js:
```diff
@@ -0,0 +1,37 @@
+});
+
+/** 이벤트가 적용되는 날짜 배열 모음 */
+export const EVENT_DAYS = Object.freeze({
+  CHRISTMAS: 25,
+  WEEKENDS: [1, 2, 8, 9, 15, 16, 22, 23, 29, 30],
+  SPECIAL: [3, 10, 17, 24, 25, 31],
+});
```
이렇게 하면 2023년 기준으로만 가능한거같아서 
아래와 같은 형식도 추천 드립니다. 

```
#getDayOfWeek(visitDate) {
	return new Date('2023-12-' + visitDate).getDay()

        /*
           visitDate는 사용자가 입력한 날짜! 
           return은 날짜에 맞는 요일이 인덱스 번호로 return 됩니다!
           Sunday - Saturday : 0 - 6
           [일,월,화,수,목,금,토] 이렇게 생각하시면 됩니다.
         */
}
```

그러면 위에 인덱스 번호로 평일 주말을 구분할수있어 매년 이렇게 상수값을 힘들게 변경하지않아도 된다고 생각해서 추천드립니다. 변경을 하면 App.js에서도 변경을 해야하지만 한번 추천드립니다! ☺️

jiyeeeah replied:
맞아요! 저 이방법 생각못했는데 태기님 코드 보면서 알았습니다 ㅎㅎ 아주 좋은 방법인 것 같아요! 👍

teagu123 reviewed (approved):
커밋은 항상 너무 잘하셔서 항상 배워갑니다!
4주차여서 그런지 다들 코드를 잘짜시는거같습니다! 또한 각자의 스타일도 좀 보이는거같구요!
깔끔하게 짜셔서 많이 배웠습니다! 수고하셨습니다

jiyeeeah committed:
[refactor] EVENT -> DISCOUNT, menuUtils분리, printDiscount 정리

- event관련 상수 더욱 상수화
  - calculate없애고, 기준이 되는 금액들만 상수화함

- 할인 혜택 계산을 위한 함수들 menuUtils로 분리
- printDiscount 정리
  - 할인 혜택 내역 없을 경우를 미리 조건문으로 파악하여 "없음"출력하고 return 하도록 수정

jiyeeeah committed:
[refactor] EVENT_BADGE 상수 전부 대문자로 수정

- 카멜 케이스에서 스네이크 케이스, 전부 대문자로 수정

jiyeeeah committed:
[refactor] OFFER_MENU 상수 전부 대문자로

jiyeeeah committed:
[refactor] MENU상수 대문자로 수정

jiyeeeah committed:
[refactor] validation 가독성 개선

- menu validation 코드 가독성 개선을 위해 함수 분리

jiyeeeah committed:
[refactor] 총 할인 금액 구하는 함수 수정

- discount 객체가 수정되면서, week안에 weekends 또는 weekdays가 있는지 확인하고 더해줘야함

jiyeeeah committed:
[refactor] return문 축약

jiyeeeah committed:
[refactor] reduce사용하여 축약

jiyeeeah committed:
[chore] 오타 수정
- baverage -> beverage

jiyeeeah committed:
[refactor] 불필요한 필드 정리

jiyeeeah committed:
[refactor] 크리스마스 디데이 할인 해당 아닐 경우 0 return하도록

jiyeeeah committed:
[bugfix] reduce 메서드 내 조건문 벗어날 경우 return 문 작성

jiyeeeah committed:
[feat] 단위 테스트 코드 추가

jiyeeeah committed:
[feat] validation 테스트 코드 추가
